File: app/src/main/java/com/lmyby/ankihelper/data/ThemeManager.java
package com.lmyby.ankihelper.data;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import androidx.core.content.ContextCompat;
import com.lmyby.ankihelper.R;

/**
 * Theme management utility class for applying themes across the app with dark mode support
 */
public class ThemeManager {
    
    /**
     * Apply theme to an activity with dark mode support
     */
    public static void applyTheme(Activity activity) {
        AppTheme theme = Settings.getInstance(activity).getSelectedTheme();
        int themeResId = theme.getThemeResId();
        
        // Apply dark mode override for themes that support system dark mode
        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(activity)) {
            themeResId = R.style.Theme_AnkiHelperDark;
        }
        
        activity.setTheme(themeResId);
    }
    
    /**
     * Apply transparent theme to popup activities with dark mode support
     */
    public static void applyTransparentTheme(Activity activity) {
        AppTheme theme = Settings.getInstance(activity).getSelectedTheme();
        int themeResId = theme.getTransparentThemeResId();
        
        // Apply dark mode override for themes that support system dark mode
        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(activity)) {
            themeResId = R.style.TransparentDark;
        }
        
        activity.setTheme(themeResId);
    }
    
    /**
     * Check if current theme is e-ink optimized
     */
    public static boolean isEinkTheme(Context context) {
        return Settings.getInstance(context).getSelectedTheme() == AppTheme.EINK;
    }
    
    /**
     * Check if current theme is dark mode
     */
    public static boolean isDarkMode(Context context) {
        AppTheme theme = Settings.getInstance(context).getSelectedTheme();
        return theme.isDarkMode() || 
               (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(context));
    }
    
    /**
     * Get theme-appropriate color resource with dark mode support
     */
    public static int getThemeColor(Context context, int defaultColorRes, 
                                   int pinkColorRes, int einkColorRes, int darkColorRes) {
        AppTheme theme = Settings.getInstance(context).getSelectedTheme();
        
        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(context)) {
            return ContextCompat.getColor(context, darkColorRes);
        }
        
        switch (theme) {
            case PINK:
                return ContextCompat.getColor(context, pinkColorRes);
            case EINK:
                return ContextCompat.getColor(context, einkColorRes);
            case DARK:
                return ContextCompat.getColor(context, darkColorRes);
            default:
                return ContextCompat.getColor(context, defaultColorRes);
        }
    }
    
    /**
     * Get theme-appropriate color resource (legacy method for backward compatibility)
     */
    public static int getThemeColor(Context context, int defaultColorRes, 
                                   int pinkColorRes, int einkColorRes) {
        return getThemeColor(context, defaultColorRes, pinkColorRes, einkColorRes, defaultColorRes);
    }
    
    /**
     * Get theme-appropriate drawable resource with dark mode support
     */
    public static int getThemeDrawable(Context context, int defaultDrawableRes,
                                      int pinkDrawableRes, int einkDrawableRes, int darkDrawableRes) {
        AppTheme theme = Settings.getInstance(context).getSelectedTheme();
        
        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(context)) {
            return darkDrawableRes;
        }
        
        switch (theme) {
            case PINK:
                return pinkDrawableRes;
            case EINK:
                return einkDrawableRes;
            case DARK:
                return darkDrawableRes;
            default:
                return defaultDrawableRes;
        }
    }
    
    /**
     * Get theme-appropriate drawable resource (legacy method for backward compatibility)
     */
    public static int getThemeDrawable(Context context, int defaultDrawableRes,
                                      int pinkDrawableRes, int einkDrawableRes) {
        return getThemeDrawable(context, defaultDrawableRes, pinkDrawableRes, einkDrawableRes, defaultDrawableRes);
    }
    
    /**
     * Restart app to apply theme changes
     */
    public static void restartAppForThemeChange(Activity activity) {
        Intent intent = activity.getPackageManager()
            .getLaunchIntentForPackage(activity.getPackageName());
        if (intent != null) {
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            activity.startActivity(intent);
            activity.finish();
            Runtime.getRuntime().exit(0);
        }
    }
}

File: app/src/main/java/com/lmyby/ankiquicker/data/ThemeManager.kt
package com.lmyby.ankiquicker.data

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import com.lmyby.ankiquicker.R

/**
 * Theme management utility class for applying themes across the app with dark mode support
 * Converted to Kotlin as part of Phase 2 data model migration
 */
object ThemeManager {

    /**
     * Apply theme to an activity with dark mode support
     */
    @JvmStatic
    fun applyTheme(activity: Activity) {
        val theme = Settings.getInstance(activity).getSelectedTheme()
        var themeResId = theme.themeResId

        // Apply dark mode override for themes that support system dark mode
        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(activity)) {
            themeResId = R.style.Theme_AnkiQuickerDark
        }

        activity.setTheme(themeResId)
    }

    /**
     * Apply transparent theme to popup activities with dark mode support
     */
    @JvmStatic
    fun applyTransparentTheme(activity: Activity) {
        val theme = Settings.getInstance(activity).getSelectedTheme()
        var themeResId = theme.transparentThemeResId

        // Apply dark mode override for themes that support system dark mode
        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(activity)) {
            themeResId = R.style.TransparentDark
        }

        activity.setTheme(themeResId)
    }

    /**
     * Check if current theme is e-ink optimized
     */
    @JvmStatic
    fun isEinkTheme(context: Context): Boolean {
        return Settings.getInstance(context).getSelectedTheme() == AppTheme.EINK
    }

    /**
     * Check if current theme is dark mode
     */
    @JvmStatic
    fun isDarkMode(context: Context): Boolean {
        val theme = Settings.getInstance(context).getSelectedTheme()
        return theme.isDarkMode() ||
                (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(context))
    }

    /**
     * Get theme-appropriate color resource with dark mode support
     */
    @JvmStatic
    @JvmOverloads
    fun getThemeColor(
        context: Context,
        defaultColorRes: Int,
        pinkColorRes: Int,
        einkColorRes: Int,
        darkColorRes: Int = defaultColorRes
    ): Int {
        val theme = Settings.getInstance(context).getSelectedTheme()

        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(context)) {
            return ContextCompat.getColor(context, darkColorRes)
        }

        return when (theme) {
            AppTheme.PINK -> ContextCompat.getColor(context, pinkColorRes)
            AppTheme.EINK -> ContextCompat.getColor(context, einkColorRes)
            AppTheme.DARK -> ContextCompat.getColor(context, darkColorRes)
            else -> ContextCompat.getColor(context, defaultColorRes)
        }
    }

    /**
     * Get theme-appropriate drawable resource with dark mode support
     */
    @JvmStatic
    @JvmOverloads
    fun getThemeDrawable(
        context: Context,
        defaultDrawableRes: Int,
        pinkDrawableRes: Int,
        einkDrawableRes: Int,
        darkDrawableRes: Int = defaultDrawableRes
    ): Int {
        val theme = Settings.getInstance(context).getSelectedTheme()

        if (theme.supportsSystemDarkMode() && theme.shouldApplyDarkMode(context)) {
            return darkDrawableRes
        }

        return when (theme) {
            AppTheme.PINK -> pinkDrawableRes
            AppTheme.EINK -> einkDrawableRes
            AppTheme.DARK -> darkDrawableRes
            else -> defaultDrawableRes
        }
    }

    /**
     * Restart app to apply theme changes
     */
    @JvmStatic
    fun restartAppForThemeChange(activity: Activity) {
        val intent = activity.packageManager.getLaunchIntentForPackage(activity.packageName)
        if (intent != null) {
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            activity.startActivity(intent)
            activity.finish()
            Runtime.getRuntime().exit(0)
        }
    }
}
